packages/filebrowser/src/model.ts

```diff
@@ -239,8 +228,7 @@ export class FileBrowserModel implements IDisposable {
    * Force a refresh of the directory contents.
    */
   refresh(): Promise<void> {
-    this._lastRefresh = new Date().getTime();
-    return this.cd('.');
+    return this._poll.refresh();
```

Comment:
You probably want to refresh, then return the poll's tick promise. The refresh just *schedules* the refresh, the actual action happens at the end of the tick.

Reply:
Or just make an async function like the tests:

```typescript
await this._poll.refresh()
await this._poll.tick
```

Reply:
Ah, thanks, I had misunderstood the function signature.

Reply:
We discussed making the refresh return the tick promise, but then it would be inconsistent with the other scheduling functions like `start`, etc. It would be convenient, though.

Reply:
We should at least make a note in the refresh method about this subtlety.

Reply:
Do you want to add the note to the poll's refresh method?

```
   *
   * #### Notes
   * The returned promise resolves after the tick is scheduled, before the
   * callback is run. To wait until after the callback executes, await the
   * `poll.tick` promise: `await poll.refresh(); await poll.tick();`
```

Reply:
Sure.

Reply:
I'm getting different results from the test from #5224 when using
```typescript
await this._poll.refresh();
await this._poll.tick;
```
compared to
```typescript
return this.cd('.');
```
It seems to me that those should be roughly equivalent (since the poll factory function is just the latter), but instead I'm getting out-of-order resolution of promises here...

Still not sure what's really going on.

Reply:
It looks like `Poll` has some logic around pending requests in which new calls to `refresh()` supersede older ones. That is not really what we want here, since calls to `FileBrowserModel.refresh()` can depend on a previous `cd` (which is what #5224 was fixing). I suspect this is a case that is not really intended to be addressed by the `Poll` functionality, where invocations of its actions are more-or-less memory-less. 

We can get around this by *not* calling `poll.refresh` in `model.refresh`, and relying on the custom logic in `cd` around pending paths, but it's not particularly clear.

Reply:
Roughly, the order I think should be (IIRC):

* Call poll.refresh
* poll.refresh calls schedule async
* poll.schedule resolves the current poll state (awaiting current tick promise resolve, which should happen after every other thing that was waiting on the current tick)
* poll.schedule then requests an animation frame to run the poll action (but does not await it).
* poll.schedule returns

At this point the await poll.refresh should return, and you await the poll.tick promise. The animation frame triggers, which awaits the poll action, then poll.schedules a 'resolved' state for the normal poll interval, and as part of the poll schedule, it resolves the poll.tick promise you were awaiting. The poll.tick promise is resolved with the poll instance, so you can check the poll state (which should be the 'resolved' state that was scheduled).

Does that help?

Reply:
so `(await poll.tick()).state.payload` should be the value returned from your factory function.

Reply:
> Roughly, the order I think should be (IIRC):
> 
>     * Call poll.refresh
This was happening as expected.
>     * poll.refresh calls schedule async
Also happening as expected.
>     * poll.schedule resolves the current poll state (awaiting current tick promise resolve, which should happen after every other thing that was waiting on the current tick)
I *don't* think this is occurring as described, in particular, the resolution of the current tick does not seem to wait on the factory function:
https://github.com/jupyterlab/jupyterlab/blob/8c72b122dcf42ac5ad02a233e84fec1032b5335a/packages/coreutils/src/poll.ts#L400-L403
>     * poll.schedule then requests an animation frame to run the poll action (but does not await it).
Agreed.
>     * poll.schedule returns
Agreed.

Reply:
> I _don't_ think this is occurring as described, in particular, the resolution of the current tick does not seem to wait on the factory function:

The poll action happens in the execute, and *after* the factory is run and the result resolves, *then* schedule is called to schedule a resolve state. That's why the tick happens after the factory runs.